### PR TITLE
RFC: index hints

### PR DESCRIPTION
**Pull request by RaduBerinde**
https://github.com/RaduBerinde/cockroach/blob/rfc-index-hints/docs/RFCS/index_hints.md

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/5762)

### CODE REVIEW

File: docs/RFCS/index_hints.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+SELECT * FROM t WITH (INDEX(myindex))
+```
+
+[4]: https://msdn.microsoft.com/en-us/library/ms187373.aspx
+
+# Alternatives
+
+We have two options:
+
+1. Change the semantics of the existing `@` syntax
```
Comment (petermattis) [inline docs/RFCS/index_hints.md]: After thinking about it more, this is my preference, perhaps extended to support `@{force_index=foo}` and `@{ignore_index=foo}`. I'm happy to just keep the current `@foo` syntax as well.

Given that everybody does these index hints differently and postgres doesn't support them at all, I don't see a strong reason to prefer one syntax over another.

**Comment by petermattis** [comment]
LGTM

**Comment by RaduBerinde** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[docs/RFCS/index_hints.md, line 110 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-KEBY4cYnHaRt5qwIGUI-r1-110:-KEBbxAHnVUFcFgQdvgj:1322772717) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e999ab29228e90940590eb604fd6eb2df74851d/docs/RFCS/index_hints.md#L110)):*
What is the difference between `@foo` and `@{force_index=foo}` ?  Are they aliases, or `@foo` = "don't use any index other than foo, i.e. foo or no index" and `@{force_index=foo}` = "use index foo"?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762)*

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[docs/RFCS/index_hints.md, line 110 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-KEBY4cYnHaRt5qwIGUI-r1-110:-KEBdXXv69EQXITKLGdg:1564690635) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e999ab29228e90940590eb604fd6eb2df74851d/docs/RFCS/index_hints.md#L110)):*
I think `@foo` and `@{force_index=foo}` should be equivalent. That is, use index `foo` regardless of the consequences. Error if index `foo` does not exist. There is merit to a slightly weaker `@{use_index=foo}` which would use index `foo` if it exists but fallback to normal index selection if it didn't. I can see utility in this behavior when adding or dropping an index from a table.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762)*

**Comment by RaduBerinde** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[docs/RFCS/index_hints.md, line 110 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-KEBY4cYnHaRt5qwIGUI-r1-110:-KEBeMHysjxEsInRrscM:-1248234221) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e999ab29228e90940590eb604fd6eb2df74851d/docs/RFCS/index_hints.md#L110)):*
Understood. I like it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762)*

**Comment by RaduBerinde** [comment]
Updated with Peter's proposal

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-:-KEBk_Yl-JQAD7_b6DUd:-35710614)*

**Comment by andreimatei** [comment]
Do we really need anything but force_index?
I don't really see the distinction with use_index being useful. There's an argument that the latter works during schema changes, but i think it's not too strong.
I also don't see ignote_index being useful. If you're providing hints, go all the way.

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-:-KEBqrpfd-QVXFOdfTxE:-1735001172)*

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/index_hints.md, line 136 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-KEBsCqNdzrYBGDPOq8o:-KEBsCqNdzrYBGDPOq8p:390648878) ([raw file](https://github.com/cockroachdb/cockroach/blob/97af4abd79b130266d693361f5bbed2e1826f621/docs/RFCS/index_hints.md#L136)):*
Sounds good to me, though I don't see any urgency behind implementing `use_index` and `ignore_index`. That is, I think we should change the semantics of `@foo`, perhaps add support for `@{force_index=foo}` and then file issues to implement `use_index` and `ignore_index` at some future time.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762)*

**Comment by RaduBerinde** [comment]
I wouldn't say there is a strong need, but they are easy enough to add - so if they can be useful to keep things working while dropping/adding indexes I don't see why not.

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-:-KEBrsoFgT8cI3fK-oUz:1074449650)*

**Comment by RaduBerinde** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[docs/RFCS/index_hints.md, line 136 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-KEBsCqNdzrYBGDPOq8o:-KEBu57lOUvGNggBoDoH:-524187984) ([raw file](https://github.com/cockroachdb/cockroach/blob/97af4abd79b130266d693361f5bbed2e1826f621/docs/RFCS/index_hints.md#L136)):*
Agreed, updated to clarify the priorities.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762)*

**Comment by andreimatei** [comment]
Well on a meta note, I think the default with modifying is to be conservative.
`ignore_index` I just don't like. What's a case where the heuristics fail you, but you don't know what index you want to use?
`use_index`... queries will rot with indexes that don't exist. And if we have `force_index`, then the user generally needs to update her queries anyway before changing these indexes. So you need to transition `force_index=old` -> `use_index=new,old` -> `force_index=new`.
I think ideally we'd only have `use_index` + some means to surface warnings to the user about indexes that don't exist, maybe in the UI. Short of that, my vote I think is just for `force_index`.

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/RFCS/index_hints.md, line 134 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-KEC24MWkvtApFL6LPbT:-KEC24MWkvtApFL6LPbU:-547646888) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d66d8d10bbb37bdbdee1bf8acd7dd237e9b4133/docs/RFCS/index_hints.md#L134)):*
what about an index that exists but doesn't cover the query, or is not sorted like the query? What about if the index truly doesn't have anything to do with the query (doesn't contain *any* of the filtering columns)?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-:-KEBushd5SgS5yIgAl6Y:1487256998)*

**Comment by andreimatei** [comment]
s//the default with modifying *languages* is to be conservative.

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-:-KEC311GGHjuHJJbip6Y:-1368041712)*

**Comment by RaduBerinde** [comment]
We may choose an index where the filter is not very effective (large spans) and results in an index-join that is too expensive. Using `ignore_index` to avoid that index has the benefit that if we later add an index that makes this query more efficient, it will get selected without going back and modifying the query.

I agree that the usecase is slim, so I'm fine with dropping these (if Peter agrees as well). I don't buy the meta argument though, we are already being conservative as these are *optional*; we are not changing any existing part of the language, we are extending it in a minor way.

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-:-KEC5a3XtljnRJAk9QWr:-981087627)*

**Comment by petermattis** [comment]
I'd hold off on `use_index` and `ignore_index` until user need reveals their utility.

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-:-KEC8JxyLHZgQQLCiEpI:899389773)*

**Comment by RaduBerinde** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/RFCS/index_hints.md, line 134 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-KEC24MWkvtApFL6LPbT:-KECAZabWt2s1UzlcU2x:-814325445) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d66d8d10bbb37bdbdee1bf8acd7dd237e9b4133/docs/RFCS/index_hints.md#L134)):*
We would still use that index - the rows will be processed in the order dictated by that index, for better or for worse. The point of the hints is that our algorithms might have missed something, so we are not in the position to determine if the choice makes sense or not. E.g. maybe there is a higher `ORDER BY` that we can't "see" because of some other node in-between (like a UNION). Something along the lines of `((SELECT .. ) UNION (SELECT. )) ORDER BY .. LIMIT 10` where using the correctly ordered index in both selects could make a *massive* difference (with the right implementation for union).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762)*

**Comment by JackKrupansky** [comment]
That's cheating, playing the "wait for 'user need'" card! Actually, the original RFC says "One example we saw recently was from the photos app, where a LIMIT clause would have made using an index a much better choice", so I would recommend going with the simplest, least invasive, easiest to implement, most compatible, but still most generally useful solution that addresses that specific need. No need to "gild the lily" unless it is simpler and faster to implement (and test... and document) that way.

**Comment by petermattis** [comment]
I disagree the waiting for "user need" is cheating. We're struggling to come up with real-world scenarios for `use_index` and `ignore_index`. Adding this functionality later will not be burdensome as it does not involve changing RPCs or data stored on disk.

**Comment by bdarnell** [comment]
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/index_hints.md, line 110 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-KEBY4cYnHaRt5qwIGUI-r1-110:-KECEwuBMedd_7xhGdVy:1028207789) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e999ab29228e90940590eb604fd6eb2df74851d/docs/RFCS/index_hints.md#L110)):*
I'm a bit worried that embedding index information in the table name will interfere with some ORMs' ability to match results up with the right table. If all existing databases with a comparable feature do it with a comment or a separate clause, that might be more supportable. Just a vague suspicion, though. 

---

*[docs/RFCS/index_hints.md, line 114 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-KECGYizaEnzspAOD9Mo:-KECGYizaEnzspAOD9Mp:-765802948) ([raw file](https://github.com/cockroachdb/cockroach/blob/df7133958b5fc06481d6f4e23b1fc34b8becda29/docs/RFCS/index_hints.md#L114)):*
If we change `@` as is currently being proposed, then a query using `foo@bar` could have one of two very different query plans depending on whether the index stores all the columns necessary (if not, it becomes like a join). I would like to see an `index-only` mode that would return an error rather than fall back to a join (this seems to me to be more useful than the use/ignore variants)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762)*

**Comment by cuongdo** [comment]
There will likely be user need for something that steers the query planner
toward specific indices. However, the details of what that is will need to
be guided by user feedback. The single `photos` app doesn't provide enough
data to properly design an addition to our SQL grammar.

**Comment by RaduBerinde** [comment]
The "main" functionality (force_index) would have been sufficient to the photos case, and we all agree we need that.

---

Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/index_hints.md, line 114 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-KECGYizaEnzspAOD9Mo:-KECQUVloMZYwOHa_2Iz:-850034) ([raw file](https://github.com/cockroachdb/cockroach/blob/df7133958b5fc06481d6f4e23b1fc34b8becda29/docs/RFCS/index_hints.md#L114)):*
The current `@` functionality is the `index-only` mode. I just don't see why it is useful. If you need a column not in the index, you need an index-join.. If you don't need columns but you are requesting them, then it's a badly written query (or suboptimally written anyway). And you can easily tell what it's doing using `EXPLAIN`. There are many ways you can shoot yourself in the foot with SQL, why should we focus on this particular angle?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-:-KECQ904HQdF8Oxn72ye:-1762924938)*

**Comment by bdarnell** [comment]
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/index_hints.md, line 114 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-KECGYizaEnzspAOD9Mo:-KECSWAtdU_g5ad1PyJ-:2145635229) ([raw file](https://github.com/cockroachdb/cockroach/blob/df7133958b5fc06481d6f4e23b1fc34b8becda29/docs/RFCS/index_hints.md#L114)):*
Index-only mode is an assertion to ensure that the data layout works the way the query author thinks it does. If someone is claiming to be able to outsmart our query planner it's good to give them this kind of tool, although it's admittedly less useful than what's being proposed for `force-index`. It's at least worth adding to the list of future extensions.

It's true that it's possible to verify the presence or absence of an index-join with `EXPLAIN`, but that tends to be a more tedious workflow than an assertion in the query itself. I found this especially frustrating when I was working with Hive HQL and seemingly minor query changes would turn a single mapreduce into a three-mapreduce pipeline, and at the time it was difficult to wire up a unittest that could verify that the query was getting EXPLAINED the right way.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762)*

**Comment by RaduBerinde** [comment]
*[docs/RFCS/index_hints.md, line 114 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-KECGYizaEnzspAOD9Mo:-KECVjM6umV02sSweQU8:-1059746535) ([raw file](https://github.com/cockroachdb/cockroach/blob/df7133958b5fc06481d6f4e23b1fc34b8becda29/docs/RFCS/index_hints.md#L114)):*
Maybe what we want then is another option that disallows use of non-covering indexes (index-joins). Used in conjunction with `force_index`, it would cause an error if the index is non-covering. Used alone, it directs the index selection to avoid all non-covering indexes. E.g. `table@foo{force_index=index,no_index_join}` and `table@foo{no_index_join}`. What do you think?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762)*

**Comment by bdarnell** [comment]
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/index_hints.md, line 114 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-KECGYizaEnzspAOD9Mo:-KECWcxlZ7bccyTzhEJq:677042770) ([raw file](https://github.com/cockroachdb/cockroach/blob/df7133958b5fc06481d6f4e23b1fc34b8becda29/docs/RFCS/index_hints.md#L114)):*
SGTM

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762)*

**Comment by RaduBerinde** [comment]
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions.

---

*[docs/RFCS/index_hints.md, line 114 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-KECGYizaEnzspAOD9Mo:-KECXYMRCEe9icAOQSBv:-495622623) ([raw file](https://github.com/cockroachdb/cockroach/blob/df7133958b5fc06481d6f4e23b1fc34b8becda29/docs/RFCS/index_hints.md#L114)):*
Cool, added to the document.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762)*

**Comment by andreimatei** [comment]
Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions.

---

*[docs/RFCS/index_hints.md, line 134 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-KECYtIpkaysgkVH94MW:-KECYtIpkaysgkVH94MX:-1846324874) ([raw file](https://github.com/cockroachdb/cockroach/blob/1db92b84027bcf03a84385454aecfef10195456f/docs/RFCS/index_hints.md#L134)):*
I think I like `no_index_join`. Do you think we should also have something for asserting the right sort order?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762)*

**Comment by tamird** [comment]
Is this unblocking something in the short term? I agree with @petermattis  that there's no urgent need to implement `use_index` and `ignore_index`, but I also feel there's not an urgent need to change the semantics of `@`.

Since we're already entering into custom syntax territory, wouldn't the current `@` syntax + explicit (if limited) JOIN support get us all the way there?

---

Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-:-KEC_IhGRu_dApt-zbLQ:71925565)*

**Comment by RaduBerinde** [comment]
IMO we should be able to provide an easy workaround when the index selection doesn't choose correctly. We have seen it ourselves once and we know there are issues with our current implementation. I don't see why we should wait for someone to report an issue when it's almost trivial to add.

Explaining to users that they have to rewrite their query to use a `JOIN` that frankly doesn't make much sense from their perspective doesn't cut it IMO. I also don't see the motivation behind it, as it would be more work to support at this point.

---

Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions.

---

*[docs/RFCS/index_hints.md, line 134 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-KECYtIpkaysgkVH94MW:-KECtNDR5unBPHmJf76n:1619406006) ([raw file](https://github.com/cockroachdb/cockroach/blob/1db92b84027bcf03a84385454aecfef10195456f/docs/RFCS/index_hints.md#L134)):*
I wouldn't. It's a lot more iffy - sometimes the order is a "nice to have" (e.g. `MIN`) - and it's a lot harder to explain. Plus as we improve our query optimization we may get better at discovering "good" orderings which would mess with existing queries.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-:-KECr_F03I_I3Ng8UNXR:-967194849)*

**Comment by RaduBerinde** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/RFCS/index_hints.md, line 110 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-KEBY4cYnHaRt5qwIGUI-r1-110:-KECvGhgxB38L_4v9Bgf:-353001470) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e999ab29228e90940590eb604fd6eb2df74851d/docs/RFCS/index_hints.md#L110)):*
@petermattis - do you have any thoughts about this?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762)*

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/RFCS/index_hints.md, line 110 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762#-KEBY4cYnHaRt5qwIGUI-r1-110:-KECvbTobSBt6yizrtFp:-1303061149) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e999ab29228e90940590eb604fd6eb2df74851d/docs/RFCS/index_hints.md#L110)):*
I don't have any knowledge of how this will mess with ORMs. I'd say leave the syntax as-is (using `@`) until we have specific knowledge of a problem.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5762)*

**Pull request merged**